pkg/webui/console/containers/latest-decoded-payload-panel/index.js

```diff
@@ -0,0 +1,197 @@
+// Copyright © 2024 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import React, { useCallback, useEffect, useState } from 'react'
+import { defineMessages } from 'react-intl'
+import { forOwn, isObject, throttle } from 'lodash'
+
+import Icon, { IconCodeDots, IconArrowsMaximize, IconX, IconCopy } from '@ttn-lw/components/icon'
+import Panel, { PanelError } from '@ttn-lw/components/panel'
+import CodeEditor from '@ttn-lw/components/code-editor'
+import Button from '@ttn-lw/components/button'
+import PortalledModal from '@ttn-lw/components/modal/portalled'
+import toast from '@ttn-lw/components/toast'
+
+import Message from '@ttn-lw/lib/components/message'
+
+import LastSeen from '@console/components/last-seen'
+
+import sharedMessages from '@ttn-lw/lib/shared-messages'
+import PropTypes from '@ttn-lw/lib/prop-types'
+
+import style from './latest-decoded-payload-panel.styl'
+
+const m = defineMessages({
+  latestDecodedPayload: 'Latest decoded payload',
+  source: 'Source: {source}',
+  received: 'Received',
+  seeInLiveData: 'See in live data',
+})
+
+const findKey = (obj, keyToFind) => {
+  let result
+
+  const recursiveSearch = obj => {
+    forOwn(obj, (value, key) => {
+      if (key === keyToFind) {
+        result = value
+        return false // Exit early
+      } else if (isObject(value)) {
+        result = recursiveSearch(value)
+        if (result) return false // Exit early if found
+      }
+    })
+    return result
+  }
+
+  return recursiveSearch(obj)
+}
+
+const LatestDecodedPayloadPanel = ({ events, shortCutLinkPath }) => {
+  const [latestEvent, setLatestEvent] = useState(null)
+  const [selectedEvent, setSelectedEvent] = useState(null)
+
+  const throttledSetLatestEvent = useCallback(
+    throttle(newEvent => setLatestEvent(newEvent), 10000),
+    [],
+  )
+
+  useEffect(() => {
+    const normalizedEvents = events
+      .map(e => ({
+        eventId: e.unique_id,
+        time: e.time,
+        decodedPayload: findKey(e.data, 'decoded_payload'),
+      }))
+      .filter(e => !!e.decodedPayload)
+    if (normalizedEvents.length > 0) {
+      throttledSetLatestEvent(normalizedEvents[0])
+    }
+  }, [events, throttledSetLatestEvent])
```

Comment:
I see what you mean but I think refactoring this `useEffect` into the code below will do the job and not be an expensive operation since it will exit with the first found event containing `decoded_payload`. I would like to avoid overcomplicating the code here. What do you think?

```
useEffect(() => {
    if (!events.length) return

    const firstEventWithPayload = events.find(e => findKey(e.data, 'decoded_payload'))

    if (firstEventWithPayload) {
      throttledSetLatestEvent({
        eventId: firstEventWithPayload.unique_id,
        time: firstEventWithPayload.time,
        decodedPayload: findKey(firstEventWithPayload.data, 'decoded_payload'),
      })
    }
  }, [events, throttledSetLatestEvent])
```